### PR TITLE
chore: Select others to receive  notifications about approved draft orders (FPLA-3227)

### PR DESCRIPTION
**Pull request by aliveni**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-3227

### Change description ###
Select others to send approved orders notifications.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: ccd-definition/AuthorisationCaseField/CareSupervision/gatekeeper.json [hunk 1]
```diff
@@ -2274,6 +2274,48 @@
     "UserRole": "caseworker-publiclaw-gatekeeper",
     "CRUD": "CRU"
   },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseFieldID": "reviewCMOOthersHeader",
+    "UserRole": "caseworker-publiclaw-gatekeeper",
+    "CRUD": "CRU"
```
Comment (toireasam) [inline ccd-definition/AuthorisationCaseField/CareSupervision/gatekeeper.json:2282]: Only a header so maybe only needs R

File: ccd-definition/AuthorisationCaseField/CareSupervision/gatekeeper.json [hunk 2]
```diff
@@ -2274,6 +2274,48 @@
...
+    "CaseFieldID": "reviewCMOShowOthers",
+    "UserRole": "caseworker-publiclaw-gatekeeper",
+    "CRUD": "CRU"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseFieldID": "reviewCMOOthersSubHeader",
+    "UserRole": "caseworker-publiclaw-gatekeeper",
+    "CRUD": "CRU"
```
Comment (toireasam) [inline ccd-definition/AuthorisationCaseField/CareSupervision/gatekeeper.json:2296]: Only a header so maybe only needs R

File: ccd-definition/AuthorisationCaseField/CareSupervision/gatekeeper.json [hunk 3]
```diff
@@ -2274,6 +2274,48 @@
...
+    "CaseFieldID": "hasOthers",
+    "UserRole": "caseworker-publiclaw-gatekeeper",
+    "CRUD": "CRU"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseFieldID": "othersToNotifyTitle",
+    "UserRole": "caseworker-publiclaw-gatekeeper",
+    "CRUD": "CRU"
```
Comment (toireasam) [inline ccd-definition/AuthorisationCaseField/CareSupervision/gatekeeper.json:2317]: Only a title so maybe only needs R

File: ccd-definition/AuthorisationCaseField/CareSupervision/magistrate.json [hunk 4]
```diff
@@ -1273,6 +1273,48 @@
     "UserRole": "caseworker-publiclaw-magistrate",
     "CRUD": "R"
   },
+  {
```
Comment (toireasam) [inline ccd-definition/AuthorisationCaseField/CareSupervision/magistrate.json:1276]: does magistrate need R perms for these if they aren't displayed in tab and they dont have access to the event?

File: ccd-definition/CaseField/CareSupervision/caseField.json [hunk 5]
```diff
@@ -1837,5 +1837,41 @@
...
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "ID": "reviewCMONotifyAllOthers",
+    "Label": "Send the orders to all others in the case?",
+    "FieldType": "YesOrNo",
+    "SecurityClassification": "Public",
+    "Searchable": "N"
+  },
```
Comment (toireasam) [inline ccd-definition/CaseField/CareSupervision/caseField.json:1867]: these fields are very similar to the one's already defined.. otherDetailsSectionSubHeader, otherDetailsSectionHeader, sendOrderToAllOthers... would be good if they could be reused but if they need to differ slightly then fine to leave. :)

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/ApproveDraftOrdersService.java [hunk 6]
```diff
@@ -262,4 +309,13 @@ private State getStateBasedOnNextHearing(CaseData caseData, ReviewDecision revie
...
     }
+
+    private String getOthersNotified(List<Element<Other>> selectedOthers) {
+        return Optional.ofNullable(selectedOthers).map(
+            others -> others.stream()
+                .filter(other -> other.getValue().isRepresented() || other.getValue()
+                    .hasAddressAdded())
+                .map(other -> other.getValue().getName()).collect(Collectors.joining(", "))
+        ).orElse(null);
+    }
```
Comment (toireasam) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/ApproveDraftOrdersService.java:320]: SealedOrderHistoryExtraOthersNotifiedGenerator contains same method, i think Soorya moved this into a more generic class that can be used by all, but maybe fine to leave here for now

File: service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/DraftOrdersApprovedEventHandlerTest.java [hunk 7]
```diff
@@ -236,21 +299,72 @@ void shouldSendOrderDocumentToRecipients() {
...
-
         verify(sendDocumentService).sendDocuments(caseData,
             List.of(hearingOrder1.getOrder(), hearingOrder2.getOrder()), List.of(representative, respondent));
-
         verifyNoMoreInteractions(sendDocumentService);
     }
 
+    @Test
+    void shouldSendOrderDocumentToRecipientsWhenServingOthersIsEnabled() {
+        final Other firstOther = Other.builder().name("other1")
```
Comment (toireasam) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/DraftOrdersApprovedEventHandlerTest.java:315]: missing wordPost is it in method name?

File: ccd-definition/AuthorisationCaseField/CareSupervision/gatekeeper.json (hunk 2 above)
Comment (aliveni) [reply ccd-definition/AuthorisationCaseField/CareSupervision/gatekeeper.json]: these fields no longer exist. I've updated the code to reuse  otherDetailsSectionSubHeader, otherDetailsSectionHeader, sendOrderToAllOthers fields

File: ccd-definition/AuthorisationCaseField/CareSupervision/gatekeeper.json (hunk 1 above)
Comment (aliveni) [reply ccd-definition/AuthorisationCaseField/CareSupervision/gatekeeper.json]: these fields no longer exist. I've updated the code to reuse otherDetailsSectionSubHeader, otherDetailsSectionHeader, sendOrderToAllOthers fields

File: ccd-definition/AuthorisationCaseField/CareSupervision/gatekeeper.json (hunk 3 above)
Comment (aliveni) [reply ccd-definition/AuthorisationCaseField/CareSupervision/gatekeeper.json:2296]: same as above, removed the fields

File: ccd-definition/AuthorisationCaseField/CareSupervision/magistrate.json (hunk 4 above)
Comment (aliveni) [reply ccd-definition/AuthorisationCaseField/CareSupervision/magistrate.json]: magistrate role doesn't have access to the event, so removed fields from this file

File: ccd-definition/CaseField/CareSupervision/caseField.json (hunk 5 above)
Comment (aliveni) [reply ccd-definition/CaseField/CareSupervision/caseField.json]: updated to reuse the existing fields

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/ApproveDraftOrdersService.java (hunk 6 above)
Comment (aliveni) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/ApproveDraftOrdersService.java:320]: it is not in master yet. I'll create a PR to cleanup before this epic is released

File: service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/DraftOrdersApprovedEventHandlerTest.java (hunk 7 above)
Comment (aliveni) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/DraftOrdersApprovedEventHandlerTest.java:315]: updated the method name

**Review by toireasam — DISMISSED** [review]
LGTM

**Review by toireasam — APPROVED** [review]

**Review by lewisbirks — APPROVED** [review]

**Pull request merged**